### ISSUE TITLE
Errors due to disallow_any_expr not always reported in incremental mode

### ISSUE THREAD

**Issue by gvanrossum** [python/mypy#4474]
I found this in an audit of internal mypy runs with and without incremental mode. It seems that when you change the `disallow_any_expr` flag, then incremental mode doesn't consider the cache dirty, so it won't recheck files that might be affected. The same may be true for other `disallow_any*` flags.

However I'm still surprised, because the code in options.py does seem to add these flags to `OPTIONS_AFFECTING_CACHE`.

**Comment by gvanrossum** [comment]
Digging a little deeper I am wondering if this may have been caused by the warnings bleeding between files. The specific run contained a diff that added this flag for two specific files and those were the files for which discrepancies were found (incremental not giving as many warnings of this nature as full mode). But *some* warnings of both forms were still reported by incremental mode:
```
Expression type contains "Any" (has type "Type[Attr]")
Expression has type "Any"
```

**Comment by gvanrossum** [comment]
Here's a repro.

- In file `gid.py`:
```py
from typing import Any

class Gid:
  def __init__(self, arg: Any) -> None: pass
```

- In file `main.py`:
```py
from gid import Gid

def f(a: object) -> None:
    isinstance(a, Gid)
```

- Do a full mypy run:
```
$ mypy --show-column-numbers --disallow-any-expr _.py 
_.py:4:18: error: Expression type contains "Any" (has type "Type[Gid]")
```

- - But an incremental mypy run is clean!
```
$ mypy -i --show-column-numbers --disallow-any-expr _.py
```

**Comment by gvanrossum** [comment]
Here's the repro for the untyped property -- it's pretty similar.

- File `ns.py`:
```py
class Namespace:
    @property
    def user(self):
        pass
```

- File `main.py`:
```py
import ns
user = ns.Namespace.user
```

- Run `mypy --show-column-numbers --disallow-any-expr main.py`:
```
main.py:2:7: error: Expression type contains "Any" (has type "Callable[[Any], Any]")
```

- Run it with `-i`, `mypy -i --show-column-numbers --disallow-any-expr main.py` gives no errors.

**Comment by gvanrossum** [comment]
Actually it hasn't got anything to do with untyped properties per se -- I can repro a similar result as in the previous comment with the following in `ns.py`:
```py
class Namespace:
    def __init__(self):
        self.user = f()
def f(): pass
```
This presents as
```
main.py:2:7: error: Expression has type "Any"
```
which is not shown with `-i`.

**Comment by gvanrossum** [comment]
Hm... With `-i` the `Any` object has its `type_of_any` attribute set to `special_form`. Without `-i` it's `unannotated` (as expected). Apparently deserializing `Any` *always* sets it to `special_form`. Fix coming up.

**Comment by ilevkivskyi** [comment]
> Apparently deserializing `Any` _always_ sets it to `special_form`. Fix coming up.

Great investigation!
